krun/vm_defs.py

```diff
@@ -480,12 +579,105 @@ def __init__(self, vm_path, env=None):
+        how we separate GC time from compilation time.
+
+        We treat the event stream like a tree. We walk this tree, keeping track
+        of the nesting of events. When computing the time for each event, we
+        exclude the time spent in nested events. The nested children have their
+        times computed separately.
+
+        The INSTRUMENTATION mapping decides which kind of events contribute
+        to which time counter (gc or compilation).
+        """
+
+        # This stores the counters for all in-process iterations
+        data = {
+            "gc_time": [],
+            "compilation_time": [],
+        }
+
+        # This stores the counters for the current in-process iteration
+        data_this_iter = {
+            "gc_time": 0,
+            "compilation_time": 0
+        }
+
+        iter_num = 0
+        current_event = None
+
+        for line in file_handle:
+            if line.startswith(INSTRUMENTATION_END_PROC_ITER_PREFIX):
+                # first some sanity checking
+                elems = line.split(":")
+                assert(len(elems) == 2 and int(elems[1]) == iter_num)
+                iter_num += 1
+                assert current_event is None
+
+                # store this iteration's data and prepare for the next
+                for event in data_this_iter.iterkeys():
+                    data[event].append(data_this_iter[event])
+                    data_this_iter[event] = 0  # reset
+                continue
+
+            # Is it the start of an event?
+            start_match = re.match(PyPyVMDef.INST_START_EVENT_REGEX, line)
+            if start_match:
+                start_time = int(start_match.groups()[0], 16)
+                event_type = start_match.groups()[1]
+
+                new_event = PyPyVMEvent(event_type, start_time, current_event)
+                current_event = new_event
+                continue
+
+            # Is it the end of an event?
+            stop_match = re.match(PyPyVMDef.INST_STOP_EVENT_REGEX, line)
+            if stop_match:
+                event_type = stop_match.groups()[1]
+                current_event.stop_time = int(stop_match.groups()[0], 16)
+
+                # check event is correctly nested
+                assert event_type == current_event.event_type
+
+                # Find which counter this event contributes to (if any)
+                for prefix, counter_name in \
+                        PyPyVMDef.INSTRUMENTATION.iteritems():
+                    if event_type.startswith(prefix):
+                        break  # found one
+                else:
+                    counter_name = None
+
+                if counter_name is not None:
+                    data_this_iter[counter_name] += current_event.get_duration()
+
+                current_event = current_event.parent
+                continue
+
+        # All events should be done and dusted
+        assert current_event is None
```

Comment:
When we enter an event, a reference to the parent event is stored inside the new event:

```
new_event = PyPyVMEvent(event_type, start_time, current_event)
```

At nesting level 0, current_event is None, so that is what is stored.

When this event is complete, we restore the old current_event value:

```
current_event = current_event.parent
```

The assertion has the effect of ensuring that we are at nesting level 0 when we finish.